### PR TITLE
JPERF-831 Migrate CI from CircleCI to Github Actions. Change the pipeline config to use IAM role and OIDC token for assuming an identity on AWS.

### PR DESCRIPTION
**Pull request by ewefie**
The reason for those changes was initially the need for migration from long-term credentials to IAM role during authentication on AWS.
Additionally, we decided to migrate the CI from CircleCI to Github Actions in one go.
After just rewriting the pipeline it turned out that not everything works as expected, I observed many failures, and as a result, the scope and time spent on the migration increased significantly.
The team decided to split the task into smaller parts, and this is the first part that consists on:
- Switching AWS auth to IAM by using the official AWS action.
- Bumping several libs:
  - Infrastructure (context: https://github.com/atlassian/infrastructure/pull/128)
  - aws-infrastructure (to change the Ubuntu image used during test execution from Xenial to Focal)
- Downgrading `virtual-users` lib to avoid failure related to https://ecosystem.atlassian.net/browse/JPERF-748
- Updating installation procedure in `CopiedDocker.kt` to avoid the error:
```
Caused by: java.util.concurrent.ExecutionException: java.util.concurrent.ExecutionException: java.lang.Exception: Error while executing sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu focal
[1201](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1202)
   stable". Exit status code SshResult(exitStatus=100, output=Hit:1 http://eu-west-1.ec2.archive.ubuntu.com/ubuntu focal InRelease
[1202](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1203)
  Hit:2 http://eu-west-1.ec2.archive.ubuntu.com/ubuntu focal-updates InRelease
[1203](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1204)
  Hit:3 http://eu-west-1.ec2.archive.ubuntu.com/ubuntu focal-backports InRelease
[1204](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1205)
  Get:4 https://download.docker.com/linux/ubuntu focal InRelease [57.7 kB]
[1205](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1206)
  Get:5 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
[1206](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1207)
  Err:4 https://download.docker.com/linux/ubuntu focal InRelease
[1207](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1208)
   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7EA0A9C3F273FCD8
[1208](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1209)
  Reading package lists...
[1209](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1210)
  , errorOutput=W: GPG error: https://download.docker.com/linux/ubuntu focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7EA0A9C3F273FCD8
[1210](https://github.com/atlassian/jira-hardware-exploration/actions/runs/3997267095/jobs/6871345394#step:8:1211)
  E: The repository 'https://download.docker.com/linux/ubuntu focal InRelease' is not signed.

```

Note: `CopiedDocker.kt` was, as the name suggests, a copy of `Docker.kt` located in `infrastructure` lib. The foremention error was already addressed in `infrastructure`, but not here. 
In this PR I followed the previous approach and just copied the implementation from `infrastructure`, but IMO keeping those two classes doesn’t make sense in a long term. I’d personally opt for getting rid of both `CopiedDocker.kt` and `CopiedDockerImage.kt` and using original classes from `infrastructure`. I’m aware though that this would require changes in API (currently they are only for internal use) - we can discuss it.

For now, tests fail and in most cases, the reason is insufficient AWS resources. It’s manifested either by error with the message: “We currently do not have sufficient xxxxx capacity in the Availability Zone you requested (eu-west-1a). Our system will be working on provisioning additional capacity.” or “ResourceStatus: CREATE_FAILED,ResourceStatusReason: Request limit exceeded.” But in general, we can see that the new auth method works, we are allowed to provision instances, and some of them are created successfully.

The most recent error I encountered is this one:
```
Caused by: java.lang.Exception: There's no max action error
[1133](https://github.com/atlassian/jira-hardware-exploration/actions/runs/4005396808/jobs/6875728893#step:6:1133)
  	at com.atlassian.performance.tools.lib.ErrorGauge.measureMaxAction(ErrorGauge.kt:20) ~[test/:?]
[1134](https://github.com/atlassian/jira-hardware-exploration/actions/runs/4005396808/jobs/6875728893#step:6:1134)
  	at com.atlassian.performance.tools.hardware.HardwareMetric.score(HardwareMetric.kt:40) ~[test/:?]
[1135](https://github.com/atlassian/jira-hardware-exploration/actions/runs/4005396808/jobs/6875728893#step:6:1135)
  	at com.atlassian.performance.tools.hardware.HardwareExploration$testHardware$1.apply(HardwareExploration.kt:296) ~[test/:?]
[1136](https://github.com/atlassian/jira-hardware-exploration/actions/runs/4005396808/jobs/6875728893#step:6:1136)
  	at com.atlassian.performance.tools.hardware.HardwareExploration$testHardware$1.apply(HardwareExploration.kt:48) ~[test/:?]
[1137](https://github.com/atlassian/jira-hardware-exploration/actions/runs/4005396808/jobs/6875728893#step:6:1137)
  	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642) ~[?:?]
[1138](https://github.com/atlassian/jira-hardware-exploration/actions/runs/4005396808/jobs/6875728893#step:6:1138)
  	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
[1139](https://github.com/atlassian/jira-hardware-exploration/actions/runs/4005396808/jobs/6875728893#step:6:1139)
  	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705) ~[?:?]
```

I found that it is caused by errors in "log in" action, but I didn't investigate it further yet.

I don’t know if there are any other issues obscured by the errors above - it is possible. I tried to run those tests in the night, I also reduced for some time the list of instances provisioned during the execution, but all I found I mentioned above or already addressed. Maybe after running those tests regularly (there’s a cron schedule for that) for a longer time some other issues appear.

cc/ @Nubzor

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Compile
+        uses: gradle/gradle-build-action@v2
+        with:
+          gradle-version: wrapper
+          arguments: compileTestKotlin
+      - name: Build
+        uses: gradle/gradle-build-action@v2
```
Comment (ewefie) [inline .github/workflows/build-and-test.yml:26]: Gradle action takes care of caches

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          role-to-assume: ${{ secrets.AWS_ROLE_TO_ASSUME }}
+          aws-region: eu-west-1
+      - name: Generate workflow URL
+        run: echo "WORKFLOW_URL=$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/actions/runs/$GITHUB_RUN_ID" >> $GITHUB_ENV
+      - name: Test integration
+        uses: gradle/gradle-build-action@v2
+        continue-on-error: true # expect it to sometimes flake
+        # timeout-minutes: 70
```
Comment (ewefie) [inline .github/workflows/build-and-test.yml]: This timeout is copied from the CircleCi - I'm not sure if we want to keep it. Wdyt?

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      branches: [ master ]
+  pull_request:
+    branches: [ master ]
+  schedule:
+    - cron: '0 0 * * 0'
+  workflow_dispatch:
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
Comment (mgrzaslewicz) [inline .github/workflows/build-and-test.yml:14]: Having mutable ubuntu version might stright us back one day, I'd keep it fixed

edit: we have fixed ubuntu version run in docker. It can stay as it is here

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (mgrzaslewicz) [reply .github/workflows/build-and-test.yml]: Having timeouts is always good. We could run it a few times, see how long it takes, add some buffer and put it here

File: src/test/kotlin/com/atlassian/performance/tools/hardware/IntegrationTestRuntime.kt [hunk 4]
```diff
@@ -16,23 +16,24 @@ object IntegrationTestRuntime {
...
-                .roleSessionName(UUID.randomUUID().toString())
-                .build(),
-            DefaultAWSCredentialsProviderChain()
-        ),
-        region = region,
-        regionsWithHousekeeping = listOf(Regions.EU_WEST_1),
-        capacity = TextCapacityMediator(region),
-        batchingCloudformationRefreshPeriod = Duration.ofSeconds(20)
-    )
+    fun prepareAws() = Aws.Builder(region)
```
Comment (mgrzaslewicz) [inline src/test/kotlin/com/atlassian/performance/tools/hardware/IntegrationTestRuntime.kt:19]: So we will spin instances without osquery. It should be addressed in another PR as soon as possible

**Review by mgrzaslewicz — DISMISSED** [review]

File: src/test/kotlin/com/atlassian/performance/tools/hardware/IntegrationTestRuntime.kt (hunk 4 above)
Comment (Nubzor) [reply src/test/kotlin/com/atlassian/performance/tools/hardware/IntegrationTestRuntime.kt:19]: This is important. @ewefie can we please have a follow-up issue in the backlog to cover this one?

**Review by pczuj — DISMISSED** [review]

File: src/test/kotlin/com/atlassian/performance/tools/hardware/IntegrationTestRuntime.kt (hunk 4 above)
Comment (ewefie) [reply src/test/kotlin/com/atlassian/performance/tools/hardware/IntegrationTestRuntime.kt:19]: Sure, here it is: https://ecosystem.atlassian.net/browse/JPERF-942

File: .github/workflows/build-and-test.yml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      contents: read
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Prepare Jira license
+        run: echo $JIRA_LICENSE > jira-license.txt
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v1
```
Comment (dagguh) [inline .github/workflows/build-and-test.yml:63]: Unfortunately, the creds aren't being refreshed during the long test:
`com.amazonaws.services.s3.model.AmazonS3Exception: The provided token has expired`
Perhaps we can tell it to use STS?

**Review by dagguh — CHANGES_REQUESTED** [review]

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (dagguh) [reply .github/workflows/build-and-test.yml:26]: Noice!

File: .github/workflows/build-and-test.yml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        with:
+          role-to-assume: ${{ secrets.AWS_ROLE_TO_ASSUME }}
+          aws-region: eu-west-1
+      - name: Generate workflow URL
+        run: echo "WORKFLOW_URL=$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/actions/runs/$GITHUB_RUN_ID" >> $GITHUB_ENV
+      - name: Test integration
+        uses: gradle/gradle-build-action@v2
+        continue-on-error: true # expect it to sometimes flake
```
Comment (dagguh) [inline .github/workflows/build-and-test.yml:72]: Nice, thanks for maintaining the previous "two runs" approach.

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (dagguh) [reply .github/workflows/build-and-test.yml]: In CircleCI it was `no_output_timeout: 70m`, so "how long can it stay silent". It was not a timeout for the whole build.
Let's see if it hangs and then add timeout as a solution to that problem.

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (ewefie) [reply .github/workflows/build-and-test.yml:63]: Yeah, the default token lifespan is one hour and we can't override this value in the action configuration. This is a limitation of this AWS action and there's even a feature request to allow credential refresh https://github.com/aws-actions/configure-aws-credentials/issues/359.
I tried to extend the session duration on AWS side, lets's see if that works.

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (ewefie) [reply .github/workflows/build-and-test.yml]: Yep, I checked if GHA has a counterpart for `no_output_timeout` but unfortunately they don't thus I was thinking about using just a regular timeout, but tbh I have no idea how long those tests can take to complete in the worst-case scenario. So yep, after some time we can decide if there's a need to introduce a timeout.


Edit: The solution doesn't work, so I think I'll need to configure AWS creds manually.

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (ewefie) [reply .github/workflows/build-and-test.yml:63]: Ok, I think I solved the issue. I extended the session duration on AWS and in the action configuration to 4 hours and there was no exception related to the expired tokens during the last run.

**Comment by dagguh** [comment]
I'm running `testIntegration` locally to debug `Caused by: java.lang.Exception: There's no max action error`.

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        with:
+          role-to-assume: ${{ secrets.AWS_ROLE_TO_ASSUME }}
+          aws-region: eu-west-1
+          role-duration-seconds: 14400
+      - name: Generate workflow URL
+        run: echo "WORKFLOW_URL=$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/actions/runs/$GITHUB_RUN_ID" >> $GITHUB_ENV
+      - name: Test integration
+        uses: gradle/gradle-build-action@v2
+        continue-on-error: true # expect it to sometimes flake
+        # timeout-minutes: 70
```
Comment (mgrzaslewicz) [inline .github/workflows/build-and-test.yml]: Leftover?

Comment (ewefie) [reply .github/workflows/build-and-test.yml]: Yeah, I missed this line. Already removed.

**Review by pczuj — APPROVED** [review]

**Review by dagguh — APPROVED** [review]
